### PR TITLE
added JTC test parameters to yaml file

### PR DESCRIPTION
**Pull request by RamCharanThota**
Hi, I have added JTC test parameters to yam file here and submitting for your review. you can find the link to respective issue here #462

### CODE REVIEW

File: joint_trajectory_controller/test/config/test_joint_trajectory_controller.yaml [hunk 1]
```diff
@@ -10,3 +10,33 @@ test_joint_trajectory_controller:
...
+  open_loop_control: true
+
+  allow_integration_in_goal_trajectories: true
+
+  action_monitor_rate: 30.0
+
+  interpolation_method: "splines"
+
+  gains:
+    __map_joints:
```
Comment (christophfroehlich) [inline joint_trajectory_controller/test/config/test_joint_trajectory_controller.yaml:29]: `__map_joints` is only a keyword for the GenerateParameterLibrary, you have to create a map for every controlled joint. E.g., see the online docs https://control.ros.org/master/doc/ros2_controllers/joint_trajectory_controller/doc/userdoc.html

File: joint_trajectory_controller/test/config/test_joint_trajectory_controller.yaml [hunk 2]
```diff
@@ -10,3 +10,33 @@ test_joint_trajectory_controller:
...
+        i: 1.0
+        d: 1.0
+        i_clamp: 10.0
+        ff_velocity_scale: 0.0
+        normalize_error: 0.02
+
+  constraints:
+     stopped_velocity_tolerance: 0.03
+     goal_time: 0.012
+     __map_joints:
```
Comment (christophfroehlich) [inline joint_trajectory_controller/test/config/test_joint_trajectory_controller.yaml:40]: same as above

File: joint_trajectory_controller/test/config/test_joint_trajectory_controller.yaml [hunk 3]
```diff
@@ -10,3 +10,33 @@ test_joint_trajectory_controller:
   state_interfaces:
     - position
     - velocity
+
+  command_joints:
```
Comment (christophfroehlich) [inline joint_trajectory_controller/test/config/test_joint_trajectory_controller.yaml:14]: Shouldn't throw this an error, if the size of `command_joints` is different than `joints`? IMHO controller chaining is not implemented for JTC yet, you might have a look if this parameter is used and what is expected there.

**Review by christophfroehlich — CHANGES_REQUESTED** [review]
Congratulations to your first PR ;) 

Please add a link to the issue you are working on in the PR description.

**Comment by christophfroehlich** [comment]
I think the purpose of the issue was also to test if the parameters were set correctly the the given non-default values (@destogl ?)
One possibility: You might add a new test in `test_trajectory_controller.cpp`, maybe called 
`TEST_P(TrajectoryControllerTestParameterized, test_non_default_parameters)` 
and see if the parameters were set correctly. `Params params_` is a protected variable and should be accessible from `std::shared_ptr<TestableJointTrajectoryController> traj_controller_`

**Comment by bmagyar** [comment]
> I think the purpose of the issue was also to test if the parameters were set correctly the the given non-default values (@destogl ?)

Be mindful of overtesting too. We don't necessarily need to test that gen params library works correctly in the controller. I think general controller behaviour with no defaults is the key here

File: joint_trajectory_controller/test/config/test_joint_trajectory_controller.yaml (hunk 1 above)
Comment (RamCharanThota) [reply joint_trajectory_controller/test/config/test_joint_trajectory_controller.yaml:29]: Thank you @christophfroehlich . I am updating it based on your suggestion. I have a question, Do I need add joints given under "command_joints" too if those joints are different from "joints". 
For example:
joints:
    - joint1
    - joint2
command_joints:
    - joint3
in this case, do i need to create map for joint1,joint2 and joint3?. I am confused between joints and command_joints. I really appreciate if you can explain a bit. Thank you

Comment (RamCharanThota) [reply joint_trajectory_controller/test/config/test_joint_trajectory_controller.yaml:29]: I defined the parameters as follows. 
```
gains:
    __map_joints:
        joint1:
          p: 10.0
          i: 1.0
          d: 1.0
          i_clamp: 10.0
          ff_velocity_scale: 0.0
          normalize_error: 0.02
        joint2:
          p: 10.0
          i: 1.0
          d: 1.0
          i_clamp: 10.0
          ff_velocity_scale: 0.0
          normalize_error: 0.02
```

is it correct format?

Comment (christophfroehlich) [reply joint_trajectory_controller/test/config/test_joint_trajectory_controller.yaml:29]: Not exactly, it works like this way: (and normalize_error is a boolean)

```
gains:
    joint1:
      p: 10.0
      i: 1.0
      d: 1.0
      i_clamp: 10.0
      ff_velocity_scale: 0.0
      normalize_error: false
    joint2:
      p: 10.0
      i: 1.0
      d: 1.0
      i_clamp: 10.0
      ff_velocity_scale: 0.0
      normalize_error: false
```

Comment (christophfroehlich) [reply joint_trajectory_controller/test/config/test_joint_trajectory_controller.yaml:29]: > Thank you @christophfroehlich . I am updating it based on your suggestion. I have a question, Do I need add joints given under "command_joints" too if those joints are different from "joints". For example: joints: - joint1 - joint2 command_joints: - joint3 in this case, do i need to create map for joint1,joint2 and joint3?. I am confused between joints and command_joints. I really appreciate if you can explain a bit. Thank you

have a look at this code here. It has to have the same size.

https://github.com/ros-controls/ros2_controllers/blob/48cc202108d4c91af2556176cae0c796f31d40fa/joint_trajectory_controller/src/joint_trajectory_controller.cpp#L559-L572

I'm not sure actually if the tests use the yaml or hardcode the parameters in every case
https://github.com/ros-controls/ros2_controllers/blob/48cc202108d4c91af2556176cae0c796f31d40fa/joint_trajectory_controller/test/test_trajectory_controller_utils.hpp#L129

If your yaml file does not throw an error: I suggest you investigating this to be sure, starting here:
https://github.com/ros-controls/ros2_controllers/blob/master/joint_trajectory_controller/test/test_trajectory_controller.cpp#L368

Comment (RamCharanThota) [reply joint_trajectory_controller/test/config/test_joint_trajectory_controller.yaml]: @christophfroehlich I went through the files you mentioned earlier. I think I have got an Idea of what's going on here. I want to double check with you whether my understand is correct or not. In this particular issue, we need to read and assign the trajectory controller param values from yaml file. We are using generate_param_library to generate header files for this parameters in yaml and we are using these param header files to retrieve and assign the values to trajectory controller parameter. And our test need to verify whether assigned trajectory controller parameters values properly matched with non default values given in the respective yaml file.

And i think you are right, we need to create new test case as you mentioned below. Also, you mentioned that I can use "param_" variable  to retrieve the parameter values using traj_controller_ object. But, I think, First, we have to  assign the values from header file to traj_controller_ using the following function in our test case
https://github.com/ros-controls/ros2_controllers/blob/48cc202108d4c91af2556176cae0c796f31d40fa/joint_trajectory_controller/test/test_trajectory_controller_utils.hpp#L195-L216

I really appreciate it if you can have look and let me whether my understand is right or wrong. If  my approach is right
1) how can i create " std::vector<rclcpp::Parameter> & parameters = {}" argument in the below function  by using yaml parameters header files created by generate param library
https://github.com/ros-controls/ros2_controllers/blob/48cc202108d4c91af2556176cae0c796f31d40fa/joint_trajectory_controller/test/test_trajectory_controller_utils.hpp#L195

2) And also how can i retrieve protected "param_" variable using traj_controller_ object when there is no getter function for that variable.

Thank you and looking forward to your reply. have a good day and sorry for long question :)

Comment (christophfroehlich) [reply joint_trajectory_controller/test/config/test_joint_trajectory_controller.yaml]: As @bmagyar has pointed out, we don't have to test if the values are read correctly from generate_parameter_library (we assume that this dependency works as expected), but test if JTC works with non-default parameters. 
Maybe @destogl can comment in the original issue which tests exactly he had in mind.
As a prerequisite for that we have to understand how we can load/use the yaml file in the tests -> I honestly also have to see and play around how this works.

**Comment by bmagyar** [comment]
I naively marked this ready for review, is that ok?

**Comment by christophfroehlich** [comment]
> I naively marked this ready for review, is that ok?

I still don't know what exactly was the idea of @destogl for the (missing?) tests and how to parse the yaml file from the tests.

@RamCharanThota The CI tests fail, because it does not even compile. Furthermore, please have a look at the [contribution guidelines](https://github.com/ros-controls/ros2_control/blob/master/CONTRIBUTING.md#contributing-via-pull-requests) on how to install pre-commit hooks for automatic formating.

**Comment by RamCharanThota** [comment]
> > I naively marked this ready for review, is that ok?
> 
> I still don't know what exactly was the idea of @destogl for the (missing?) tests and how to parse the yaml file from the tests.
> 
> @RamCharanThota The CI tests fail, because it does not even compile. Furthermore, please have a look at the [contribution guidelines](https://github.com/ros-controls/ros2_control/blob/master/CONTRIBUTING.md#contributing-via-pull-requests) on how to install pre-commit hooks for automatic formating.

I already installed the pre-commit hooks. But, I will double check. Thank you

**Comment by RamCharanThota** [comment]
> > I naively marked this ready for review, is that ok?
> 
> I still don't know what exactly was the idea of @destogl for the (missing?) tests and how to parse the yaml file from the tests.
> 
> @RamCharanThota The CI tests fail, because it does not even compile. Furthermore, please have a look at the [contribution guidelines](https://github.com/ros-controls/ros2_control/blob/master/CONTRIBUTING.md#contributing-via-pull-requests) on how to install pre-commit hooks for automatic formating.

Whenever I tried to commit. I am getting the following error

amet_cpp_error

any idea how can i solve it. I am not doing anything complex. I am just declaring a function
function_declaration

**Comment by christophfroehlich** [comment]
I always use `export AMENT_CPPCHECK_ALLOW_SLOW_VERSIONS=true` to get that running.

File: joint_trajectory_controller/test/config/test_joint_trajectory_controller.yaml (hunk 1 above)
Comment (RamCharanThota) [reply joint_trajectory_controller/test/config/test_joint_trajectory_controller.yaml]: Hi @christophfroehlich , I am still struggling to understand what needs to be done exactly in this test and how to retrieve the parameters. I wondering whether you get the chance to have a look at it?

**Comment by RamCharanThota** [comment]
> This pull request is in conflict. Could you fix it @RamCharanThota?

Hi @mergify

**Pull request closed without merging**